### ISSUE TITLE
Affinity/Trust system (Backend)

### ISSUE THREAD

**Issue by ghost** [Just-Natsuki-Team/NatsukiModDev#12]
# Affinity and Trust System
### Trust:
- -100 - 100 scale
- The player should *start* at a negative value, as ideally Natsuki wouldn't be very trusting of you.
  - This can be mitigated if the player followed Natsuki's routes in DDLC
- Trust increases along with affinity, as trust and affinity are directly correlated
- Trust also decreases along with affinity for the same reason

### Affinity:
- Effectively acts as a relationship-progres-scale
- Leads to *intimacy*, another subsystem which is begun once a certain stage
    - Intimacy is basically the more *risque*-ish aspects of the relationship. Where Natsuki's swimsuit changes, and she's willing to say more risque things
    - This should be gated by an age check

### Correlation of Trust and Affinity:
- Trust and Affinity are both linked internally, at least somewhat.
    - As trust increases, affinity should too, vice-versa
    - As affinity increases, trust should too, vice-versa

|Trust Level|Affinity Level|Behaviour|
|:----------:|:----------:|:----------:|
|Low|Low|Natsuki pretty much hates you. She cannot trust you to do to do anything nor treat her right (no better than her dad)|
|Medium|Low|Natsuki trusts you a little bit, but she still pretty much hates you|
|High|Low|Shouldn't be possible. Considering she hates you, her trust shouldn't be able to be high|
|Low|Medium|Natsuki doesn't trust you at all, but she's developing feelings for you|
|Medium|Medium|This can be considered an *early* phase in the relationship. She's starting to trust you and is developing feelings for you. She's still kinda defensive about her feelings for you, though|
|High|Medium|Natsuki really trusts you, but and she does like you, but she's not completely in love with you yet. There's still further to go in the relationship|
|Low|High|Shouldn't be possible, trust is a limiting factor in the relationship and wouldn't let it progress this far|
|Medium|High|Natsuki really likes you, but she'd still be defensive about it since she doesn't fully trust you yet. But she might be more open with you about some things|
|High|High|Natsuki trusts you completely and she loves you whole-heartedly. She's more than happy to completely open up to you.|

## TODO:
- [ ] Build a model for both affinity and trust separately
- [x] Build functions that can increment/decrement the trust and affinity values
- [x] Crossbuild them so they have the appropriate impact on each other
- [ ] Develop constants to represent different trust and love states
- [x] Develop functions to evaluate different states (check if we're at and/or above/below states) as a result of this system (potential vector system to measure a reasonable response?)
- [ ] Add logging and audit functions to track gains/losses
- [ ] Figure out a timescale to transition from state-to-state and develop the mathematical model to represent each subsystem

## Things to note:
- This system should NEVER be used via a standard `+=` or `-=` to increment/decrement either trust/affinity
- We should ALWAYS log changes to affinity and trust. This way we can validate that things work properly
- We may need to make a cap for how much trust/affinity we can earn for a time period, that way we can make it impossible to "speedrun" the relationship

**Comment by spaghetsie** [comment]
I'm not sure if this contradicts with not using `+=` `-=` or not, but I was thinking of having:

**Trust increase (doesn't matter the scale, ReLU takes care of it)**
`Trust += 1+ ReLU(Affinity/20)`
This way we would have a base 1 increase no matter how low **affinity** is and having high **affinity** would make it easier to develop **trust**.

**Trust decrease (if affinity is -100 - 100 scale)**
`Trust -= 6 + abs(Affinity-50)/10`
At very low **affinity** loosing trust is easier. At high **affinity** Natsuki would feel betrayed making you loose more **trust**

**Trust decrease (if affinity is 0 - 100 scale)**
`Trust -= 11 - (Affinity-50)*2/10`
Same as above but converting `0 - 100` to `-100 - 100`

**Affinity increase**
`Affinity += ReLU(Trust/20)`
Having negative **trust** makes it impossible to progress in a relationship.

**Affinity decrease**
`Affinity -= 5 - ReLU((Trust-50)/25)`
High **trust** would make Natsuki think it was just some kind of a missunderstanding. **trust** at or below 50 wouldn't affect Affinity decrease.


The values can be tampered with to slow/speed up progression. This is just a basic concept I though of.

**Comment by Blizzardsev** [comment]
Closing, as this has been superseded by updates.
